src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+        # execute pulse simulation in emulator
+        ro_reduced_dm, rdm_qubit_list = self.simulation_backend.qevolve(
+            channel_waveforms, self.simulate_dissipation
+        )
+
+        return ro_reduced_dm, rdm_qubit_list
+
+    def play(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers: Dict[QubitId, Coupler],
+        sequence: PulseSequence,
+        execution_parameters: Optional[ExecutionParameters] = None,
+    ) -> dict[str, Union[IntegratedResults, SampleResults]]:
+        """Executes the sequence of instructions and generates readout results.
+
+        Args:
+            qubits (dict): Qubits involved in the device. Does not affect emulator.
+            couplers (dict): Couplers involved in the device. Does not affect emulator.
+            sequence (`qibolab.pulses.PulseSequence`): Pulse sequence to simulate.
+            execution_parameters (`qibolab.ExecutionParameters`): Parameters (nshots,
+                                                        relaxation_time,
+                                                        fast_reset,
+                                                        acquisition_type,
+                                                        averaging_mode)
+                                                        Defaults to None, for which case it uses the values stored in self.exec_params.
+
+        Returns:
+            dict: A dictionary mapping the readout pulses serial and respective qubits to
+            Qibolab results object.
+        """
+        if execution_parameters is None:
+            execution_parameters = self.exec_params
+
+        nshots = execution_parameters.nshots
+        ro_pulse_list = sequence.ro_pulses
+        ro_reduced_dm, rdm_qubit_list = self.run_pulse_simulation(
+            sequence, self.instant_measurement
+        )
+        samples = self.get_samples(nshots, ro_reduced_dm, rdm_qubit_list)
+        results = self.get_results_from_samples(
+            ro_pulse_list, samples, execution_parameters
+        )
+
+        return results
+
+    def print_sim_details(self, sim_index=-1):
+        """Print simulation details of any of the previously run simulations.
+
+        Defaults to -1, i.e. the last simulation.
+        """
+        full_time_list = self.simulation_backend.pulse_sim_time_list[sim_index]
+        print("Hamiltonian:", self.simulation_backend.H[sim_index])
+        print("Initial state:", self.simulation_backend.psi0)
+        # print("Full time list:", full_time_list)
+        print("Initial simualtion time:", full_time_list[0])
+        print("Final simualtion time:", full_time_list[-1])
+        print("Simualtion time step (dt):", full_time_list[1])
+        print("Total number of time steps:", len(full_time_list))
+        print("Static dissipators:", self.simulation_backend.static_dissipators)
+        print("Simulation options:", self.simulation_backend.sim_opts)
```

Comment:
Don't print, return a dictionary

Reply:
Removed in d8b2ebd32e5d0bc392ee02eab54875ac842500ab, added essential ones to results.